File: apps/web/src/types/site.ts
export type InfrastructureIntent =
  | "GENERAL_AI_INFRASTRUCTURE"
  | "DATA_CENTER_FEASIBILITY"
  | "PUBLIC_COMPUTE_HUB"
  | "EDGE_AI_NODES"
  | "CLOUD_FIRST_STRATEGY"
  | "FIBER_CONNECTIVITY_UPGRADE"
  | "POWER_GRID_READINESS"
  | "CITY_DATA_PLATFORM"
  | "AI_LITERACY_PROGRAM"
  | "GOVERNANCE_CYBERSECURITY"
  | "SECTOR_SPECIFIC_READINESS";

export type LegacyInfrastructureType =
  | "PUBLIC_AI_COMPUTE_HUB"
  | "REGIONAL_AI_DATA_CENTER";

export type InfrastructureType =
  | InfrastructureIntent
  | LegacyInfrastructureType;

export type ScenarioType =
  | "BUILD_NOW"
  | "UPGRADE_FIBER_FIRST"
  | "VALIDATE_GRID_FIRST"
  | "AI_LITERACY_TRAINING"
  | "CLOUD_FIRST"
  | "DELAY_INVESTMENT"
  | "GOVERNANCE_FIRST"
  | "EDGE_PILOT_FIRST"
  | "OPEN_DATA_PLATFORM_FIRST";

export type LayerKey =
  | "education"
  | "healthcare"
  | "government"
  | "overall_readiness";

export type ConfidenceLevel = "Low" | "Medium" | "High";

export type ReviewStatus =
  | "DRAFT_ANALYSIS"
  | "NEEDS_MORE_DATA"
  | "READY_FOR_EXPERT_REVIEW"
  | "REVIEWED_FOR_PLANNING_ONLY"
  | "ESCALATED_TO_AUTHORITY";

export type EvidenceDecision =
  | "UNVERIFIED"
  | "NEEDS_SOURCE"
  | "REVIEWED"
  | "REJECTED"
  | "REQUIRES_EXPERT_VALIDATION";

export interface SelectedLocation {
  lat: number;
  lng: number;
  label?: string;
}

export interface AnalyzeSitePayload {
  lat: number;
  lng: number;
  intent?: InfrastructureIntent;
  infrastructureIntent?: InfrastructureIntent;
  infrastructureType?: InfrastructureType;
  userQuestion?: string;
  activeLayers: string[];
  scenario: ScenarioType;
}

export interface Suitability {
  score: number;
  level: string;
  confidence: ConfidenceLevel;
  recommendation: string;
}

export interface ComponentScores {
  power: number;
  connectivity: number;
  coolingWater: number;
  physicalFeasibility: number;
  computeEcosystem: number;
  sectorDemand: number;
  governance: number;
  digitalAccess: number;
  aiLiteracy: number;
  dataMaturity: number;
  equity: number;
  resilience: number;
  dataCompleteness: number;
  dataFreshness: number;
  sourceReliability: number;
  geographicResolution: number;
}

export interface SectorReadiness {
  name: string;
  level: string;
  score: number;
  mainGap: string;
  suggestedUseCases: string[];
}

export interface RoadmapItem {
  horizon: string;
  actions: string[];
}

export interface ScoreFormulaTerm {
  component: string;
  label: string;
  weightPercent: number;
  score: number;
  contribution: number;
  direction: "readiness" | "gap" | "priority" | "context";
  explanation: string;
}

export interface ScenarioImpact {
  component: string;
  label: string;
  beforeScore: number;
  afterScore: number;
  delta: number;
  explanation: string;
}

export interface PlanningContext {
  focusLabel: string;
  focusQuestion: string;
  scenarioLabel: string;
  scenarioDescription: string;
  scoreFormula: ScoreFormulaTerm[];
  relevantComponents: string[];
  scenarioImpacts: ScenarioImpact[];
  focusSpecificEvidenceNeeds: string[];
  focusSpecificWarnings: string[];
}

export interface EvidenceSummary {
  activeLayerCount: number;
  scoredLayerCount: number;
  realOpenLayerCount: number;
  syntheticLayerCount: number;
  matchedFeatureCount: number;
  nearestEvidenceKm: number | null;
  summary: string;
  confidenceImpact: string;
}

export interface ScoreDriver {
  component: string;
  score: number;
  evidenceCount: number;
  nearestEvidenceKm: number | null;
  supportingLayers: string[];
  openDataSupportingLayers: string[];
  syntheticSupportingLayers: string[];
  excludedSyntheticLayers: string[];
  explanation: string;
  includedInFocusScore: boolean;
  formulaWeight: number | null;
  scenarioAdjustment: number;
  focusSpecificExplanation: string;
}

export interface MatchedEvidence {
  layerId: string;
  layerLabel: string;
  name: string;
  assetType: string;
  category: string;
  source: string;
  sourceType: string;
  sourceConfidence: string;
  dataCompleteness: string;
  geometryType: string;
  distanceKm: number | null;
  relation: string;
  dataLimitation: string;
}

export interface ExcludedSyntheticLayer {
  layerId: string;
  layerLabel: string;
  reason: string;
}

export interface AgentReview {
  summary: string;
  scoreReliability: string;
  evidenceStrengths: string[];
  evidenceGaps: string[];
  uncertaintyNotes: string[];
  challengedAssumptions: string[];
  nextValidationSteps: string[];
  evidenceCitations: string[];
  scoreDriverSummary: string;
  excludedEvidenceNotes: string[];
  usedLlm: boolean;
}

export interface ScoreExplanation {
  headline: string;
  strongestDrivers: string[];
  weakestDrivers: string[];
  dataQualityBadge: string;
  dataQualitySummary: string;
}

export interface AgentTrace {
  intentSource: string;
  classifier: string;
  classifierConfidence: string;
  classifierReason: string;
  toolsUsed: string[];
  activeLayerCount: number;
  scoredLayerCount: number;
  openDataLayerCount: number;
  syntheticDemoLayerCount: number;
  guardrailsTriggered: number;
}

export interface GapSummary {
  digitalAccess: string;
  aiLiteracy: string;
  infrastructure: string;
  dataQuality: string;
}

export interface SiteAnalysisResult {
  intent: InfrastructureIntent;
  querySummary: string;
  selectedSite: {
    lat: number;
    lng: number;
    label: string;
  };
  suitability: Suitability;
  componentScores: ComponentScores;
  sectors: SectorReadiness[];
  bottlenecks: string[];
  strengths: string[];
  priorityInvestments: string[];
  roadmap: RoadmapItem[];
  planningContext: PlanningContext;
  evidenceSummary: EvidenceSummary;
  scoreDrivers: ScoreDriver[];
  matchedEvidence: MatchedEvidence[];
  excludedSyntheticLayers: ExcludedSyntheticLayer[];
  dataGaps: string[];
  scoreExplanation: ScoreExplanation;
  agentTrace: AgentTrace;
  agentReview: AgentReview;
  confidenceExplanation: string;
  gapSummary: GapSummary;
  recommendedInfrastructurePath: string;
  humanReviewRequired: boolean;
  warnings: string[];
}

export interface ReviewChecklistItem {
  id: string;
  category: string;
  label: string;
  description: string;
  checked: boolean;
  notes: string;
}

export interface EvidenceReviewItem {
  id: string;
  evidenceId: string;
  layerId: string;
  layerLabel: string;
  name: string;
  sourceType: string;
  sourceConfidence: string;
  dataCompleteness: string;
  dataLimitation: string;
  decision: EvidenceDecision;
  notes: string;
}

export interface ReviewerNote {
  id: string;
  author: string;
  note: string;
  createdAt: string;
}

export interface HumanReviewRecord {
  reviewId: string;
  createdAt: string;
  updatedAt: string;
  status: ReviewStatus;
  selectedSite: SiteAnalysisResult["selectedSite"];
  planningFocus: string;
  scenario: string;
  suitabilityScore: number;
  confidence: string;
  evidenceSummary: EvidenceSummary;
  matchedEvidenceIds: string[];
  syntheticDemoEvidenceCount: number;
  reviewerNotes: ReviewerNote[];
  checklistItems: ReviewChecklistItem[];
  evidenceItems: EvidenceReviewItem[];
  requiredValidationSteps: string[];
  sourceDisclosure: string;
  warnings: string[];
  nonGoalWarning: string;
}

export interface ReviewPacket extends HumanReviewRecord {
  generatedAt: string;
}

export interface ChatMessage {
  id: string;
  role: "user" | "assistant";
  content: string;
}

export interface AgentChatContext {
  selectedLocation: SelectedLocation | null;
  analysis: SiteAnalysisResult | null;
  review: HumanReviewRecord | null;
  activeLayers: string[];
  scenario: ScenarioType;
  planningFocus: InfrastructureIntent;
}

export interface CandidateZone {
  id: string;
  label: string;
  lat: number;
  lng: number;
  description: string;
}

File: apps/api/app/schemas/site.py
"""Pydantic contracts shared by site analysis and agent chat endpoints."""

from __future__ import annotations

from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, Field


class InfrastructureIntent(StrEnum):
    GENERAL_AI_INFRASTRUCTURE = "GENERAL_AI_INFRASTRUCTURE"
    DATA_CENTER_FEASIBILITY = "DATA_CENTER_FEASIBILITY"
    PUBLIC_COMPUTE_HUB = "PUBLIC_COMPUTE_HUB"
    EDGE_AI_NODES = "EDGE_AI_NODES"
    CLOUD_FIRST_STRATEGY = "CLOUD_FIRST_STRATEGY"
    FIBER_CONNECTIVITY_UPGRADE = "FIBER_CONNECTIVITY_UPGRADE"
    POWER_GRID_READINESS = "POWER_GRID_READINESS"
    CITY_DATA_PLATFORM = "CITY_DATA_PLATFORM"
    AI_LITERACY_PROGRAM = "AI_LITERACY_PROGRAM"
    GOVERNANCE_CYBERSECURITY = "GOVERNANCE_CYBERSECURITY"
    SECTOR_SPECIFIC_READINESS = "SECTOR_SPECIFIC_READINESS"


class ScenarioType(StrEnum):
    BUILD_NOW = "BUILD_NOW"
    UPGRADE_FIBER_FIRST = "UPGRADE_FIBER_FIRST"
    VALIDATE_GRID_FIRST = "VALIDATE_GRID_FIRST"
    AI_LITERACY_TRAINING = "AI_LITERACY_TRAINING"
    CLOUD_FIRST = "CLOUD_FIRST"
    DELAY_INVESTMENT = "DELAY_INVESTMENT"
    GOVERNANCE_FIRST = "GOVERNANCE_FIRST"
    EDGE_PILOT_FIRST = "EDGE_PILOT_FIRST"
    OPEN_DATA_PLATFORM_FIRST = "OPEN_DATA_PLATFORM_FIRST"


ConfidenceLevel = Literal["Low", "Medium", "High"]


class SelectedSite(BaseModel):
    lat: float
    lng: float
    label: str = "Selected AI planning site"


class AnalyzeSiteRequest(BaseModel):
    lat: float
    lng: float
    intent: str | None = None
    infrastructureIntent: str | None = None
    infrastructureType: str | None = None
    userQuestion: str | None = None
    activeLayers: list[str] = Field(default_factory=list)
    scenario: ScenarioType = ScenarioType.BUILD_NOW


class Suitability(BaseModel):
    score: int
    level: str
    confidence: ConfidenceLevel
    recommendation: str


class ComponentScores(BaseModel):
    power: int
    connectivity: int
    coolingWater: int
    physicalFeasibility: int
    computeEcosystem: int
    sectorDemand: int
    governance: int
    digitalAccess: int
    aiLiteracy: int
    dataMaturity: int
    equity: int
    resilience: int
    dataCompleteness: int
    dataFreshness: int
    sourceReliability: int
    geographicResolution: int


class SectorReadiness(BaseModel):
    name: str
    level: str
    score: int
    mainGap: str
    suggestedUseCases: list[str]


class RoadmapItem(BaseModel):
    horizon: str
    actions: list[str]


class ScoreFormulaTerm(BaseModel):
    component: str
    label: str
    weightPercent: int
    score: int
    contribution: float
    direction: Literal["readiness", "gap", "priority", "context"] = "readiness"
    explanation: str = ""


class ScenarioImpact(BaseModel):
    component: str
    label: str
    beforeScore: int
    afterScore: int
    delta: int
    explanation: str


class PlanningContext(BaseModel):
    focusLabel: str
    focusQuestion: str
    scenarioLabel: str
    scenarioDescription: str
    scoreFormula: list[ScoreFormulaTerm]
    relevantComponents: list[str]
    scenarioImpacts: list[ScenarioImpact]
    focusSpecificEvidenceNeeds: list[str]
    focusSpecificWarnings: list[str]


class EvidenceSummary(BaseModel):
    activeLayerCount: int
    scoredLayerCount: int
    realOpenLayerCount: int
    syntheticLayerCount: int
    matchedFeatureCount: int
    nearestEvidenceKm: float | None = None
    summary: str
    confidenceImpact: str


class ScoreDriver(BaseModel):
    component: str
    score: int
    evidenceCount: int
    nearestEvidenceKm: float | None = None
    supportingLayers: list[str]
    openDataSupportingLayers: list[str] = Field(default_factory=list)
    syntheticSupportingLayers: list[str] = Field(default_factory=list)
    excludedSyntheticLayers: list[str]
    explanation: str
    includedInFocusScore: bool = False
    formulaWeight: int | None = None
    scenarioAdjustment: int = 0
    focusSpecificExplanation: str = ""


class MatchedEvidence(BaseModel):
    layerId: str
    layerLabel: str
    name: str
    assetType: str
    category: str
    source: str
    sourceType: str
    sourceConfidence: str
    dataCompleteness: str
    geometryType: str
    distanceKm: float | None = None
    relation: str
    dataLimitation: str


class ExcludedSyntheticLayer(BaseModel):
    layerId: str
    layerLabel: str
    reason: str


class AgentReview(BaseModel):
    summary: str
    scoreReliability: str
    evidenceStrengths: list[str]
    evidenceGaps: list[str]
    uncertaintyNotes: list[str]
    challengedAssumptions: list[str]
    nextValidationSteps: list[str]
    evidenceCitations: list[str] = Field(default_factory=list)
    scoreDriverSummary: str = ""
    excludedEvidenceNotes: list[str] = Field(default_factory=list)
    usedLlm: bool = False


class ScoreExplanation(BaseModel):
    headline: str
    strongestDrivers: list[str]
    weakestDrivers: list[str]
    dataQualityBadge: str
    dataQualitySummary: str


class AgentTrace(BaseModel):
    intentSource: str
    classifier: str
    classifierConfidence: str
    classifierReason: str
    toolsUsed: list[str]
    activeLayerCount: int
    scoredLayerCount: int
    openDataLayerCount: int
    syntheticDemoLayerCount: int
    guardrailsTriggered: int


class GapSummary(BaseModel):
    digitalAccess: str
    aiLiteracy: str
    infrastructure: str
    dataQuality: str


class SiteAnalysisResult(BaseModel):
    intent: InfrastructureIntent
    querySummary: str
    selectedSite: SelectedSite
    suitability: Suitability
    componentScores: ComponentScores
    sectors: list[SectorReadiness]
    bottlenecks: list[str]
    strengths: list[str]
    priorityInvestments: list[str]
    roadmap: list[RoadmapItem]
    planningContext: PlanningContext
    evidenceSummary: EvidenceSummary
    scoreDrivers: list[ScoreDriver]
    matchedEvidence: list[MatchedEvidence]
    excludedSyntheticLayers: list[ExcludedSyntheticLayer]
    dataGaps: list[str]
    scoreExplanation: ScoreExplanation
    agentTrace: AgentTrace
    agentReview: AgentReview
    confidenceExplanation: str
    gapSummary: GapSummary
    recommendedInfrastructurePath: str
    humanReviewRequired: bool
    warnings: list[str]


class ChatMessage(BaseModel):
    id: str
    role: Literal["user", "assistant"]
    content: str


class ChatRequest(BaseModel):
    message: str
    currentAnalysis: SiteAnalysisResult | None = None
    currentReview: dict[str, Any] | None = None
    hasSelectedLocation: bool = False
    selectedLocation: SelectedSite | None = None
    activeLayers: list[str] = Field(default_factory=list)
    scenario: ScenarioType = ScenarioType.BUILD_NOW
    planningFocus: str | None = None
